pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,179 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclaim
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"regexp"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/samber/lo"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const tagPeriod = 1 * time.Second
+
+type Controller struct {
+	kubeClient    client.Client
+	ec2           *ec2.EC2
+	defaultTagRgx *regexp.Regexp
+	timer         *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		ec2:           ec2.New(sess),
+		defaultTagRgx: regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:         time.NewTimer(tagPeriod),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Extract resource ID from node spec
+	resourceID := path.Base(nodeClaim.Status.ProviderID)
+	if resourceID == "." || resourceID == "/" {
+		return reconcile.Result{}, fmt.Errorf("failed to extract resource ID from providerID")
+	}
+
+	// Check if instance has already been tagged
+	if isTagged, err := c.isInstanceTagged(ctx, nodeClaim, resourceID); err != nil || isTagged {
+		return reconcile.Result{}, err
+	}
+
+	// Tag instance
+	err := c.tagInstance(ctx, nodeClaim, resourceID)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+	}
+
+	logging.FromContext(ctx).
+		With("node", req.NamespacedName.Name).
+		With("instanceID", resourceID).
+		Debugf("updated ec2 instance Name tag")
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(onlyTaggableInstances()),
+	)
+}
+
+func onlyTaggableInstances() predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(object client.Object) bool {
+		nodeClaim := object.(*corev1beta1.NodeClaim)
+
+		// Instance has already been tagged
+		if val, ok := nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged]; ok && val == "true" {
```

Comment:
This doesn't work in all cases where there is already another update event that has made it through the filter while the event is being processed. There should be a check after the `Get` in the reconcile that validates this same condition as the check on the event filter to reduce the calls as much as possible